CLN: Moving Series.rank and DataFrame.rank to generic.py

nbonnotte opened the pull request:
I'm pulling this out from PR #11918. It is a first step toward solving issue #11759

The signature of `Series.rank` changes, and is now the same as `DataFrame.rank`. This may cause some problems...

jreback commented on pandas/core/series.py:
```diff
@@ -1842,35 +1842,6 @@ def argsort(self, axis=0, kind='quicksort', order=None):
                 dtype='int64').__finalize__(self)
 
-    def rank(self, method='average', na_option='keep', ascending=True,
-             pct=False):
```
I think that we should have the signature:

``def rank(self, method='average', axis=0, numeric_only=None, na_option='keep', ascending=True, pct=False)``

this will be more consistent with other methods.

nbonnotte replied:
I've modified the signature

jreback commented:
need a whatsnew entry to explain this change (can be pretty short, just demonstrating the old/new)

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -247,6 +247,8 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
    ser
    pd.Timestamp('2012-01-01') - ser
 
+- ``Series.rank()``, ``DataFrame.rank()`` and their groupby equivalents have a new, common signature ``rank(self, method='average', axis=0, numeric_only=None, na_option='keep', ascending=True, pct=False)``. (:issue:`11759`)
```
can you add this in a sub-section and show usage (of the changes, mainly to Series).

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,80 @@ def last(self, offset):
+
+        # numeric_only does not make sense for a Series
+        from pandas import Series
+        if numeric_only and isinstance(self, Series):
```
use ``ABCSeries``

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,80 @@ def last(self, offset):
+        from pandas import Series
+        if numeric_only and isinstance(self, Series):
+            msg = "numeric_only parameter not valid for Series"
+            raise NotImplementedError(msg)
```
this need to test ``not None``

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,80 @@ def last(self, offset):
+
+        # if numeric_only is None, and we can't get anything, we try with
+        # numeric_only=True
+        # this does not make sense for a Series
```
take out 3796 comment (as we are by definition not a Series)

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,80 @@ def last(self, offset):
+
+        # Series does not have columns, and its constructor does not accept
+        # a columns parameter
+        if hasattr(data, "columns"):
```
use 

```
ranks = self._constructor(ranks, **self._construct_axes_dict())
```

nbonnotte commented:
Is it ok like that?

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,73 @@ def last(self, offset):
+
+        # this method does not make sense when ndim > 2
+        if self.ndim > 2:
+            raise NotImplementedError
```
give a nice message

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,73 @@ def last(self, offset):
+                raise NotImplementedError(msg)
+        # if numeric_only is None, and we can't get anything, we try with
+        # numeric_only=True
+        elif not isinstance(self, Series):
```
use ``ABCSeries`` (as you did above)

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,73 @@ def last(self, offset):
+        from pandas import Series
+
+        # numeric_only does not make sense for a Series
+        if numeric_only is not None:
```
I would actually leave out the entire biz of checking if this is a ``Series``. If numeric only is selected it will work if its numeric and otherwise return an empty Series.

nbonnotte commented:
I've changed the error message, and removed the check for `Series`.

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -248,6 +248,35 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The **new** signature is ``.rank(method='average', axis=0, numeric_only=None, na_option='keep', ascending=True, pct=False)``.
+
```
@jorisvandenbossche what do you think?

might be more useful to have ``df.rank(1)`` allowed (and use the original ``DataFrame`` signature``, as probably more useful then specifying the ``method``

jreback replied:
and this is more consistent with things like ``df.sum(1)``

nbonnotte replied:
What about `Series.rank` then? If `axis` comes first, then it means the first parameter is the least useful...

jreback replied:
hmm, this is what I did for sorting, so precedent for what you have here as well.

```
In [1]: Series.sort_values?
Type:        instancemethod
String form: <unbound method Series.sort_values>
File:        c:\miniconda\envs\neat\lib\site-packages\pandas\core\series.py
Definition:  Series.sort_values(self, axis=0, ascending=True, inplace=False, kind='quicksort', na_position='last')
Docstring:
Sort by the values along either axis

.. versionadded:: 0.17.0

Parameters
----------
by : string name or list of names which refer to the axis items
axis : index to direct sorting
ascending : bool or list of bool
     Sort ascending vs. descending. Specify list for multiple sort orders.
     If this is a list of bools, must match the length of the by
inplace : bool
     if True, perform operation in-place
kind : {`quicksort`, `mergesort`, `heapsort`}
     Choice of sorting algorithm. See also ndarray.np.sort for more information.
     `mergesort` is the only stable algorithm. For DataFrames, this option is
     only applied when sorting on a single column or label.
na_position : {'first', 'last'}
     `first` puts NaNs at the beginning, `last` puts NaNs at the end

Returns
-------
sorted_obj : Series
```

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -248,6 +248,35 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
    pd.Timestamp('2012-01-01') - ser
 
+
+Signature change for Series.rank and DataFrame.rank and their GroupBy equivalents
```
Signature change for ``.rank``

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -248,6 +248,35 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
+Signature change for Series.rank and DataFrame.rank and their GroupBy equivalents
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The **new** signature is ``.rank(method='average', axis=0, numeric_only=None, na_option='keep', ascending=True, pct=False)``.
```
use separate code-blocks here:

Previous Signature

New Signature

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,65 @@ def last(self, offset):
+        numeric_only : boolean, default None
+            Include only float, int, boolean data. Valid only for DataFrame or
+            Panel objects
+        method : {'average', 'min', 'max', 'first', 'dense'}
```
the signature is good, but reorg the doc-string to match

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,65 @@ def last(self, offset):
+
+        Parameters
+        ----------
+        axis : int or string, default 0
```
```
axis: int or string, default 0
   index to direct ranking
```

nbonnotte replied:
What do you mean?

jreback commented on pandas/core/generic.py:
```diff
@@ -3747,6 +3748,65 @@ def last(self, offset):
+        # numeric_only=True
+        if numeric_only is None:
+            try:
+                ranks = algos.rank(self.values, axis=axis, method=method,
```
define the function so you don't have to repeat it, e.g.

```
ranker = lambda data: algos.rank(data, axis=axis, method=method,ascending=ascending, na_option=na_option,
                                    pct=pct)
```

then call it
``ranks = ranker(self.values)``
and below
``ranks = ranker(data.values)``

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -248,6 +248,35 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
+
+Signature change for Series.rank and DataFrame.rank and their GroupBy equivalents
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
mark this as issue #11759

jreback commented:
can you rebase/update

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -3751,6 +3752,66 @@ def last(self, offset):
+
+        Parameters
+        ----------
+        axis: {index (0), columns (1)), default 0
```
can you use the same type description as before in the dataframe docstring? (`{0 or 'index', 1 or 'columns'}, default 0`) -> quotes to indicate it is a string argument

nbonnotte commented:
@jorisvandenbossche i've changed the docstring as you suggested, and added a test to check if `.rank(axis=0)` and `.rank(axis='index')` on the one hand, and `.rank(axis=1)` and `.rank(axis='columns')` on the other hand, are equivalent.

nbonnotte commented:
Uh?

```
ERROR: test_get_underlying_price (pandas.io.tests.test_data.TestYahooOptions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/util/testing.py", line 1831, in wrapper
    return t(*args, **kwargs)
  File "/home/travis/build/pydata/pandas/pandas/io/tests/test_data.py", line 402, in test_get_underlying_price
    quote_price = options_object._underlying_price_from_root(root)
  File "/home/travis/build/pydata/pandas/pandas/io/data.py", line 739, in _underlying_price_from_root
    underlying_price = root.xpath('.//*[@class="time_rtq_ticker Fz-30 Fw-b"]')[0]\
IndexError: list index out of range
```

I don't understand.

The same error appeared in the pull request #11582. Is it me, or is this another issue?

jreback commented:
neh, just back network connectivity with Travis, ignore / try again

nbonnotte commented:
It's my lucky day
```
======================================================================
ERROR: test_get_goog_volume (pandas.io.tests.test_data.TestGoogle)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/indexes/base.py", line 1900, in get_value
    return tslib.get_value_box(s, key)
  File "pandas/tslib.pyx", line 807, in pandas.tslib.get_value_box (pandas/tslib.c:17638)
    cpdef object get_value_box(ndarray arr, object loc):
  File "pandas/tslib.pyx", line 816, in pandas.tslib.get_value_box (pandas/tslib.c:17306)
    i = <Py_ssize_t> loc
TypeError: 'str' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/tseries/index.py", line 1344, in get_value
    return _maybe_box(self, Index.get_value(self, series, key),
  File "/home/travis/build/pydata/pandas/pandas/indexes/base.py", line 1908, in get_value
    raise e1
  File "/home/travis/build/pydata/pandas/pandas/indexes/base.py", line 1894, in get_value
    return self._engine.get_value(s, k)
  File "pandas/index.pyx", line 103, in pandas.index.IndexEngine.get_value (pandas/index.c:3204)
    cpdef get_value(self, ndarray arr, object key):
  File "pandas/index.pyx", line 111, in pandas.index.IndexEngine.get_value (pandas/index.c:2903)
    loc = self.get_loc(key)
  File "pandas/index.pyx", line 575, in pandas.index.DatetimeEngine.get_loc (pandas/index.c:10977)
    self._date_check_type(val)
  File "pandas/index.pyx", line 581, in pandas.index.DatetimeEngine._date_check_type (pandas/index.c:11145)
    raise KeyError(val)
KeyError: 'JAN-02-2015'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pandas/index.pyx", line 565, in pandas.index.DatetimeEngine.get_loc (pandas/index.c:10766)
    return self.mapping.get_item(val.value)
  File "pandas/hashtable.pyx", line 303, in pandas.hashtable.Int64HashTable.get_item (pandas/hashtable.c:6527)
    cpdef get_item(self, int64_t val):
  File "pandas/hashtable.pyx", line 309, in pandas.hashtable.Int64HashTable.get_item (pandas/hashtable.c:6465)
    raise KeyError(val)
KeyError: 1420156800000000000

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/tseries/index.py", line 1354, in get_value
    return self.get_value_maybe_box(series, key)
  File "/home/travis/build/pydata/pandas/pandas/tseries/index.py", line 1364, in get_value_maybe_box
    values = self._engine.get_value(_values_from_object(series), key)
  File "pandas/index.pyx", line 103, in pandas.index.IndexEngine.get_value (pandas/index.c:3204)
    cpdef get_value(self, ndarray arr, object key):
  File "pandas/index.pyx", line 111, in pandas.index.IndexEngine.get_value (pandas/index.c:2903)
    loc = self.get_loc(key)
  File "pandas/index.pyx", line 567, in pandas.index.DatetimeEngine.get_loc (pandas/index.c:10832)
    raise KeyError(val)
KeyError: Timestamp('2015-01-02 00:00:00')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/util/testing.py", line 1831, in wrapper
    return t(*args, **kwargs)
  File "/home/travis/build/pydata/pandas/pandas/io/tests/test_data.py", line 93, in test_get_goog_volume
    self.assertEqual(df.Volume.ix['JAN-02-2015'], 1446662)
  File "/home/travis/build/pydata/pandas/pandas/core/indexing.py", line 77, in __getitem__
    return self._getitem_axis(key, axis=0)
  File "/home/travis/build/pydata/pandas/pandas/core/indexing.py", line 993, in _getitem_axis
    return self._get_label(key, axis=axis)
  File "/home/travis/build/pydata/pandas/pandas/core/indexing.py", line 88, in _get_label
    return self.obj[label]
  File "/home/travis/build/pydata/pandas/pandas/core/series.py", line 558, in __getitem__
    result = self.index.get_value(self, key)
  File "/home/travis/build/pydata/pandas/pandas/tseries/index.py", line 1356, in get_value
    raise KeyError(key)
KeyError: 'JAN-02-2015'
----------------------------------------------------------------------
Ran 9254 tests in 337.827s
FAILED (SKIP=152, errors=1)
```

nbonnotte commented:
I must have jumped in a parallel universe at some point, the tests used to have failed...

@jreback all green now

jorisvandenbossche commented:
sorry, no parallel universe, I just restarted the test .. :-) 
Maybe time to remove those data tests

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -282,6 +282,36 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
+     1    2
+     dtype: float64
+
+     In [4]: pd.DataFrame([0,1]).rank(axis=0, numeric_only=None, method='average', na_option='keep', ascending=True, pct=False)
```
wrap this so it lines up nicely ([3]) as well

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -282,6 +282,36 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
+.. ipython:: python
+
+     pd.Series([0,1]).rank(axis=0, method='average', numeric_only=None, na_option='keep', ascending=True, pct=False)
+     pd.DataFrame([0,1]).rank(axis=0, method='average', numeric_only=None, na_option='keep', ascending=True, pct=False)
```
same here

jreback commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -875,6 +875,12 @@ def test_rank_na_option(self):
+        assert_frame_equal(df.rank(axis=0), df.rank(axis='index'))
+        assert_frame_equal(df.rank(axis=1), df.rank(axis='columns'))
+
     def test_sem(self):
```
can you add onto the rank tests in ``tests/series/test_analytics.py`` simulating the original signature (and asserting an error), e.g.
someone tries. ``s.rank('average')`` e.g. w/o the kw. just for a back-compat that it raises the correct kind of error

nbonnotte replied:
Done

jreback commented:
minor comments. ping when pushed and green.

nbonnotte committed:
CLN: Moving Series.rank and DataFrame.rank to generic.py

nbonnotte commented:
All green

jreback commented:
thanks!
